core/src/main/java/org/infinispan/conflict/impl/StateReceiverImpl.java

```diff
@@ -0,0 +1,233 @@
+      if (stateChunks.isEmpty()) {
+         if (trace)
+            log.tracef("Ignoring received state for cache %s from %s because stateChunks are empty", cacheName, sender);
+         return;
+      }
+
+      int segmentId = stateChunks.iterator().next().getSegmentId();
+      SegmentRequest request = requestMap.get(segmentId);
+      if (request == null) {
+         if (trace) log.tracef("Ignoring received state for cache %s because the associated request was completed or cancelled %s", cacheName);
+         return;
+      }
+      request.receiveState(sender, topologyId, stateChunks);
+   }
+
+   Map<K, Map<Address, InternalCacheEntry<K, V>>> getKeyReplicaMap(int segmentId) {
+      return requestMap.get(segmentId).keyReplicaMap;
+   }
+
+   Map<Address, InboundTransferTask> getTransferTaskMap(int segmentId) {
+      return requestMap.get(segmentId).transferTaskMap;
+   }
+
+   InboundTransferTask createTransferTask(int segmentId, Address source, CacheTopology topology) {
+      return new InboundTransferTask(Collections.singleton(segmentId), source, topology.getTopologyId(),
+            rpcManager, commandsFactory, transferTimeout, cacheName, false);
+   }
+
+   class SegmentRequest {
+      final int segmentId;
+      final LocalizedCacheTopology topology;
+      final List<Address> replicaHosts;
+      final Map<K, Map<Address, InternalCacheEntry<K, V>>> keyReplicaMap = new HashMap<>();
+      final Map<Address, InboundTransferTask> transferTaskMap = new HashMap<>();
+      volatile CompletableFuture<List<Map<Address, InternalCacheEntry<K, V>>>> future;
+
+      SegmentRequest(int segmentId, LocalizedCacheTopology topology) {
+         this.segmentId = segmentId;
+         this.topology = topology;
+         this.replicaHosts = topology.getDistributionForSegment(segmentId).writeOwners();
+      }
+
+      synchronized CompletableFuture<List<Map<Address, InternalCacheEntry<K, V>>>> requestState() {
+         assert future == null;
+         if (trace) log.tracef("Attempting to receive replicas for segment %s from %s with topology %s", segmentId, replicaHosts, topology);
+
+         List<CompletableFuture<Void>> completableFutures = new ArrayList<>();
+         for (Address replica : replicaHosts) {
+            if (replica.equals(rpcManager.getAddress())) {
+               dataContainer.forEach(entry -> {
+                  int keySegment = topology.getDistribution(entry.getKey()).segmentId();
+                  if (keySegment == segmentId) {
+                     addKeyToReplicaMap(replica, entry);
+                  }
+               });
+            } else {
+               InboundTransferTask transferTask = createTransferTask(segmentId, replica, topology);
+               transferTaskMap.put(replica, transferTask);
+               completableFutures.add(transferTask.requestSegments());
+            }
+         }
+         future = CompletableFuture
```

Comment:
I would also think about limiting the degree of parallelism here. I recall issues in 32 node tests, where a node joining the cluster requested state from all other nodes at once - and the other 31 nodes started sending it, effectively DDoSing the poor joiner.